core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -930,6 +962,69 @@ public void onFailure(Exception shardFailedError) {
         }
     }
 
+    /** a wrapper class to encapsulate a request when being sent to a specific allocation id **/
+    final class RequestWithAllocationID<R extends TransportRequest> extends TransportRequest {
+
+        /** {@link AllocationId#getId()} of the shard this request is sent to **/
+        private String allocationId;
+
+        private R request;
+
+        RequestWithAllocationID(Supplier<R> requestSupplier) {
+            request = requestSupplier.get();
+            allocationId = null;
+        }
+
+        RequestWithAllocationID(R request, String allocationId) {
```

Comment:
I guess this constructor parameter should be `targetAllocationID` too to be consistent with my other suggestions.